lib/vssh/libssh2.c

```diff
@@ -680,28 +680,40 @@ static CURLcode ssh_force_knownhost_key_type(struct connectdata *conn)
   static const char * const hostkey_method_ssh_dss
     = "ssh-dss";
 
-
   const char *hostkey_method = NULL;
   struct ssh_conn *sshc = &conn->proto.sshc;
   struct Curl_easy *data = conn->data;
   struct libssh2_knownhost* store = NULL;
+  const char *kh_name_end = NULL;
+  long unsigned int kh_name_size = 0;
+  int port = 0;
   bool found = false;
 
-  /*&& data->set.str[STRING_SSH_KNOWNHOSTS]
-   *<-- unnecessary because of sshc->kh? */
   if(sshc->kh && !data->set.str[STRING_SSH_HOST_PUBLIC_KEY_MD5]) {
     /* lets try to find our host in the known hosts file */
     while(!libssh2_knownhost_get(sshc->kh, &store, store)) {
-      /* TODO: what about specific ports, mentioned in the name? */
-      if(strcmp(store->name, conn->host.name) == 0) {
-        infof(data, "Found host (%s) in %s\n",
-            conn->host.name, data->set.str[STRING_SSH_KNOWNHOSTS]);
+      /* For non-standard ports, the name will be enclosed in */
+      /* square brackets, followed by a colon and the port */
+      if(store->name[0] == '[') {
+        kh_name_end = strstr(store->name, "]:");
+        port = atoi(kh_name_end + 2);
+        if(kh_name_end && (port == conn->remote_port)) {
+          kh_name_size = strlen(store->name) - 1 - strlen(kh_name_end);
+          if(strncmp(store->name + 1, conn->host.name, kh_name_size) == 0) {
+            found = true;
+            break;
+          }
+        }
+      }
```

Comment:
I think this would be better preserved in libssh2 so that the libssh2_knownhost struct has an additional port member.